golang/cosmos/app/app.go

```diff
@@ -803,14 +804,88 @@ func NewAgoricApp(
 	return app
 }
 
+type swingStoreMigrationEventHandler struct {
+	swingStore sdk.KVStore
+}
+
+func (eventHandler swingStoreMigrationEventHandler) OnExportStarted(height uint64, retrieveSwingStoreExport func() error) error {
+	return retrieveSwingStoreExport()
+}
+
+func (eventHandler swingStoreMigrationEventHandler) OnExportRetrieved(provider swingsetkeeper.SwingStoreExportProvider) (err error) {
+	defer helpers.RecoverToError(&err)
+
+	var exportDataIterator sdk.Iterator
+	exportDataIterator, err = provider.GetExportDataIterator()
+	if err != nil {
+		return err
+	}
+	defer exportDataIterator.Close()
+
+	if !exportDataIterator.Valid() {
+		return fmt.Errorf("swing store export must provide some export data")
+	}
+
+	for ; exportDataIterator.Valid(); exportDataIterator.Next() {
+		eventHandler.swingStore.Set(exportDataIterator.Key(), exportDataIterator.Value())
+	}
+	return nil
+}
+
 // upgrade11Handler performs standard upgrade actions plus custom actions for upgrade-11.
 func upgrade11Handler(app *GaiaApp, targetUpgrade string) func(sdk.Context, upgradetypes.Plan, module.VersionMap) (module.VersionMap, error) {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVm module.VersionMap) (module.VersionMap, error) {
 		app.CheckControllerInited(false)
 		// Record the plan to send to SwingSet
 		app.upgradePlan = &plan
 
-		// TODO: Migrate x/vstorage swingStore to x/swingset SwingStore
+		savedBlockHeight := uint64(ctx.BlockHeight() - 1)
+
+		// First, repair swing-store metadata in case this node was state-synced
+		// This is done without checks on the block height as it's idempotent,
+		// and only missing historical metadata is imported
+		getExportDataIterator := func() (iter sdk.Iterator, err error) {
```

Comment:
Suggested clarifying rename:
```suggestion
		getSwingStoreDataFromVstorage := func() (iter sdk.Iterator, err error) {
```
(or some variation thereof)